plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlScanner.java

```diff
@@ -34,12 +67,74 @@ public void startElement(String uri, String localName, String qName, Attributes
         || "http://java.sun.com/xml/ns/javaee".equals(uri))) {
       String version = attributes.getValue("version");
       if (!version.equals("2.5")) {
+    	// Adds <web-app> element to the queue if the Servlet version is not 2.5.
         Locator2 locator = getLocator();
         DocumentLocation start = new DocumentLocation(locator.getLineNumber(),
             locator.getColumnNumber());
         addToBlacklist(new JavaServletElement(Messages.getString("web.xml.version"), start, 0));
       }
     }
+    if ("servlet-class".equals(localName)) {
+      Locator2 locator = getLocator();
+      insideServletClassElement = true;
+      servletClassElementContents = new StringBuilder();
+      servletClassElementLocation =
+          new DocumentLocation(locator.getLineNumber(), locator.getColumnNumber());
+    }
+  }
+  
+  @Override
+  public void characters(char ch[], int start, int length) throws SAXException {
+    if (insideServletClassElement) {
+      servletClassElementContents.append(ch, start, length);
+    }
+  }
+  
+  @Override
+  public void endElement(String uri, String localName, String qName) throws SAXException {
+    if ("servlet-class".equals(localName)) {
+      insideServletClassElement = false;
+      String servletClassName = servletClassElementContents.toString();
+      if (!classExists(project, servletClassName)) {
+    	// Adds <servlet-class> element to the queue if the class doesn't exist in the project.
+        addToBlacklist(new UndefinedServletElement(
+            servletClassName, servletClassElementLocation, servletClassName.length()));
+      }
+    }
+  }
+  
+  @VisibleForTesting
+  static boolean classExists(IJavaProject project, String typeName) {
+    if (Strings.isNullOrEmpty(typeName)) {
+      return false;
+    }
+    SearchPattern pattern = SearchPattern.createPattern(typeName, 
+        IJavaSearchConstants.CLASS,
+        IJavaSearchConstants.DECLARATIONS,
+        SearchPattern.R_EXACT_MATCH | SearchPattern.R_ERASURE_MATCH);
+    IJavaSearchScope scope = project == null ? SearchEngine.createWorkspaceScope()
+        : SearchEngine.createJavaSearchScope(new IJavaElement[] {project});
+    return performSearch(pattern, scope, null);
+  }
+  
+  /**
+   * Searches for a class that matches a pattern.
+   */
+  @VisibleForTesting
+  static boolean performSearch(SearchPattern pattern, IJavaSearchScope scope,
+      IProgressMonitor monitor) {
+    try {
+      SearchEngine searchEngine = new SearchEngine();
+      ClassSearchRequestor requestor = new ClassSearchRequestor();
+      searchEngine.search(pattern,
+          new SearchParticipant[] { SearchEngine.getDefaultSearchParticipant() },
+          scope, requestor, monitor);
+      List<SearchMatch> results = requestor.getResults();
```

Comment:
Is it worth accumulating the results if you're not actually using it?  `CountingSearchRequestor` with `count()`?